app/actions.py

```diff
@@ -560,37 +560,64 @@ def restoreUserHistory(self):
     initialize the variables to default values.
 
     Args:
-      None
+      None.
 
     Returns:
-      None
+      None.
     """
-    # Restore the file history
+    # Restore the file history.
     self.fileHistory = app.history.getFileHistory(self.fullPath, self.data)
-    # Restore cursor position.
+
+    # Restore all positions and values of variables.
     self.view.cursorRow, self.view.cursorCol = self.fileHistory.setdefault(
         'cursor', (0, 0))
     self.penRow, self.penCol = self.fileHistory.setdefault('pen', (0, 0))
-    # self.view.scrollRow, self.view.scrollCol = self.fileHistory.setdefault('scroll', (0, 0))
-    # Determine optimal cursor position.
-    oRow = int(app.prefs.editor['optimalCursorRow'] * (self.view.rows - 3))
-    self.view.scrollRow = max(0, min(len(self.lines) - 1, self.penRow - oRow))
-    oCol = int(app.prefs.editor['optimalCursorCol'] * (self.view.cols - 1))
-    self.view.scrollCol = max(0, self.penCol - oCol)
-    # Restore selection.
+    self.view.scrollRow, self.view.scrollCol =  self.fileHistory.setdefault(
+        'scroll', (0, 0))
+    self.view.scrollRow, self.view.scrollCol = self.optimalScrollPosition(
+        self.penRow, self.penCol)
     self.doSelectionMode(self.fileHistory.setdefault('selectionMode',
         app.selectable.kSelectionNone))
     self.markerRow, self.markerCol = self.fileHistory.setdefault('marker',
         (0, 0))
-    # Restore redo chain.
     self.redoChain = self.fileHistory.setdefault('redoChain', [])
-    # Restore indices.
     self.savedAtRedoIndex = self.fileHistory.setdefault('savedAtRedoIndex', 0)
     self.redoIndex = self.savedAtRedoIndex
+
     # Store the file's info.
     self.lastChecksum, self.lastFileSize = app.history.getFileInfo(
         self.fullPath)
 
+  def optimalScrollPosition(self, row, col):
+    """
+    Calculates the optimal position for the view.
+
+    Args:
+      row (int): The cursor's row position.
+      col (int): The cursor's column position.
+
+    Returns:
+      A tuple of (row, col) representing where
+      the view should be placed.
+    """
+    optimalRowRatio = app.prefs.editor['optimalCursorRow']
+    optimalColRatio = app.prefs.editor['optimalCursorCol'] 
+    maxRows = self.view.rows
+    maxCols = self.view.cols
+    scrollRow = self.view.scrollRow
+    scrollCol = self.view.scrollCol
+    if not (scrollRow <= row <= scrollRow + maxRows and
+        scrollCol <= col <= scrollCol + maxCols):
```

Comment:
This if statement is supposed to check if the cursor is inside the view. Imagine when **scrollRow** is at 0, and **maxRows** = 25. Then row can be at most 25 since the 25th row is still in the view. The same argument goes for scrollCol, so I think they're both <=. The reason I don't just do > is because I think it improves the understanding of the code if it's written like this.